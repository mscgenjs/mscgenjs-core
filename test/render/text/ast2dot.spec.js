const fs       = require("fs");
const path     = require("path");
const renderer = require("../../../dist/render/text/ast2dot").default;
const fix      = require("../../astfixtures.json");

describe('render/text/ast2dot', () => {
    describe('#renderAST() - mscgen classic compatible - simple syntax trees', () => {

        test("should, given a simple syntax tree, render a dot script", () => {
            expect(renderer.render(fix.astSimple)).toMatchSnapshot();
        });

        test("should, given a syntax tree with boxes, render a dot script", () => {
            expect(renderer.render(fix.astBoxArcs)).toMatchSnapshot();
        });

    });

    describe('#renderAST() - xu compatible', () => {
        test('alt only - render correct script', () => {
            expect(renderer.render(fix.astOneAlt)).toMatchSnapshot();
        });
        test('alt within loop - render correct script', () => {
            expect(renderer.render(fix.astAltWithinLoop)).toMatchSnapshot();
        });
    });

    describe('#renderAST() - file based tests', () => {
        test('should render all arcs', () => {
            const lASTString = fs.readFileSync(
                path.join(__dirname, "../../fixtures/test01_all_possible_arcs_mscgen.json"),
                {"encoding":"utf8"}
            );
            const lAST = JSON.parse(lASTString);
            expect(renderer.render(lAST)).toMatchSnapshot();
        });
    });
});
