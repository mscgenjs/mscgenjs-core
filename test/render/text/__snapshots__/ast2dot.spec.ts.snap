// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render/text/ast2dot #renderAST() - file based tests should render all arcs 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]

  "0" [label="0"];
  "a" [label="a"];
  "b" [label="b"];
  "c" [label="c"];
  "d" [label="d"];
  "e" [label="e"];
  "f" [label="f"];
  "g" [label="g"];
  "h" [label="h"];
  "i" [label="i"];
  "j" [label="j"];
  "k" [label="k"];

  "a" -- "b" [label="(1) left to right", dir="none"]
  "b" -- "c" [label="(2) right to left", dir="none"]
  "c" -- "d" [label="(3) left to right
bi-directional

", dir="none"]
  "d" -- "e" [label="(4) right to left
bi-directional

", dir="none"]
  "e" -- "f" [label="(5) left to right
non-directional

", dir="none"]
  "f" -- "g" [label="(6) left to right
non-directional

", dir="none"]
  "g" -- "h" [label="(7) left to right
self reference

", dir="none"]
  "h" -- "i" [label="(8) right to left
self reference

", dir="none"]
  "i" -- "j" [label="(9) bi-directional
self reference

", dir="none"]
  "j" -- "k" [label="(10) non-directional
self reference

", dir="none"]
  "0" -- "a" [label="(11) signal", dir="none"]
  "a" -- "b" [label="(12) a -> b", arrowhead="rvee"]
  "c" -- "b" [label="(13) b <- c", arrowhead="rvee"]
  "c" -- "d" [label="(14) c <-> d", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "e" -- "d" [label="(15) e <-> d", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "e" -- "f" [label="(16) e -- f", dir="none"]
  "g" -- "f" [label="(17) g -- f", dir="none"]
  "g" -- "g" [label="(18) g -> g", arrowhead="rvee"]
  "h" -- "h" [label="(19) h <- h", arrowhead="rvee"]
  "i" -- "i" [label="(20) i <-> i", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "j" -- "j" [label="(21) j -- j", dir="none"]
  "0" -- "a" [label="(22) method", dir="none"]
  "a" -- "b" [label="(23) a => b", arrowhead="normal"]
  "c" -- "b" [label="(24) b <= c", arrowhead="normal"]
  "c" -- "d" [label="(25) c <=> d", arrowhead="normal", arrowtail="normal", dir="both"]
  "e" -- "d" [label="(26) e <=> d", arrowhead="normal", arrowtail="normal", dir="both"]
  "e" -- "f" [label="(27) e == f", dir="none"]
  "g" -- "f" [label="(28) g == f", dir="none"]
  "g" -- "g" [label="(29) g => g", arrowhead="normal"]
  "h" -- "h" [label="(30) h <= h", arrowhead="normal"]
  "i" -- "i" [label="(31) i <=> i", arrowhead="normal", arrowtail="normal", dir="both"]
  "j" -- "j" [label="(32) j == j", dir="none"]
  "0" -- "a" [label="(33) return", dir="none"]
  "a" -- "b" [label="(34) a >> b", style="dashed"]
  "c" -- "b" [label="(35) b << c", style="dashed"]
  "c" -- "d" [label="(36) c <<>> d", style="dashed", dir="both"]
  "e" -- "d" [label="(37) e <<>> d", style="dashed", dir="both"]
  "e" -- "f" [label="(38) e .. f", style="dashed", dir="none"]
  "g" -- "f" [label="(39) g .. g", style="dashed", dir="none"]
  "g" -- "g" [label="(40) g >> g", style="dashed"]
  "h" -- "h" [label="(41) h << h", style="dashed"]
  "i" -- "i" [label="(42) i <<>> i", style="dashed", dir="both"]
  "j" -- "j" [label="(43) j .. j", style="dashed", dir="none"]
  "0" -- "a" [label="(44) callback", dir="none"]
  "a" -- "b" [label="(45) a =>> b"]
  "c" -- "b" [label="(46) b <<= c"]
  "c" -- "d" [label="(47) c <<=>> d", dir="both"]
  "e" -- "d" [label="(48) e <<=>> d", dir="both"]
  "g" -- "g" [label="(49) g =>> g"]
  "h" -- "h" [label="(50) g <<= g"]
  "i" -- "i" [label="(51) i <<=>>i ", dir="both"]
  "0" -- "a" [label="(52) lost", dir="none"]
  "a" -- "b" [label="(53) a -x b", arrowhead="oinvonormal"]
  "c" -- "b" [label="(54) b x- c", arrowhead="oinvonormal"]
  "g" -- "g" [label="(55) g -x g", arrowhead="oinvonormal"]
  "h" -- "h" [label="(56) h x- h", arrowhead="oinvonormal"]
  "0" -- "a" [label="(57) emphasised", dir="none"]
  "a" -- "b" [label="(58) a :> b", style="bold"]
  "c" -- "b" [label="(59) b <: c", style="bold"]
  "c" -- "d" [label="(60) c <:> d", style="bold", dir="both"]
  "e" -- "d" [label="(61) e <:> d", style="bold", dir="both"]
  "e" -- "f" [label="(62) e :: f", style="bold", dir="none"]
  "g" -- "f" [label="(63) g :: f", style="bold", dir="none"]
  "g" -- "g" [label="(64) g :> g", style="bold"]
  "h" -- "h" [label="(65) h <: h", style="bold"]
  "i" -- "i" [label="(66) i <:> i", style="bold", dir="both"]
  "j" -- "j" [label="(67) j :: j", style="bold", dir="none"]
  "0" -- "a" [label="(68) note", dir="none"]
  box69 [label="b note c", shape="note"]
  box69 -- {"b","c"} [style="dotted", dir="none"]
  box70 [label="e note d", shape="note"]
  box70 -- {"e","d"} [style="dotted", dir="none"]
  box71 [label="g note g", shape="note"]
  box71 -- {"g","g"} [style="dotted", dir="none"]
  "0" -- "a" [label="(72) box", dir="none"]
  box73 [label="b box c", shape="box"]
  box73 -- {"b","c"} [style="dotted", dir="none"]
  box74 [label="e box d", shape="box"]
  box74 -- {"e","d"} [style="dotted", dir="none"]
  box75 [label="g box g", shape="box"]
  box75 -- {"g","g"} [style="dotted", dir="none"]
  "0" -- "a" [label="(76) rbox", dir="none"]
  box77 [label="b rbox c", style="rounded", shape="box"]
  box77 -- {"b","c"} [style="dotted", dir="none"]
  box78 [label="e rbox d", style="rounded", shape="box"]
  box78 -- {"e","d"} [style="dotted", dir="none"]
  box79 [label="g rbox g", style="rounded", shape="box"]
  box79 -- {"g","g"} [style="dotted", dir="none"]
  "0" -- "a" [label="(80) abox", dir="none"]
  box81 [label="b abox c", shape="hexagon"]
  box81 -- {"b","c"} [style="dotted", dir="none"]
  box82 [label="e abox d", shape="hexagon"]
  box82 -- {"e","d"} [style="dotted", dir="none"]
  box83 [label="g abox g", shape="hexagon"]
  box83 -- {"g","g"} [style="dotted", dir="none"]
  "0" -- "a" [label="(84) empty row", dir="none"]
  "0" -- "a" [label="(85) comment", dir="none"]
  "0" -- "a" [label="(86) broadcasts", dir="none"]
  "a" -- "0" [label="(87) a -> *", arrowhead="rvee"]
  "a" -- "b" [label="(88) a -> *", arrowhead="rvee"]
  "a" -- "c" [label="(89) a -> *", arrowhead="rvee"]
  "a" -- "d" [label="(90) a -> *", arrowhead="rvee"]
  "a" -- "e" [label="(91) a -> *", arrowhead="rvee"]
  "a" -- "f" [label="(92) a -> *", arrowhead="rvee"]
  "a" -- "g" [label="(93) a -> *", arrowhead="rvee"]
  "a" -- "h" [label="(94) a -> *", arrowhead="rvee"]
  "a" -- "i" [label="(95) a -> *", arrowhead="rvee"]
  "a" -- "j" [label="(96) a -> *", arrowhead="rvee"]
  "a" -- "k" [label="(97) a -> *", arrowhead="rvee"]
  "c" -- "0" [label="(98) * <- c", arrowhead="rvee"]
  "c" -- "a" [label="(99) * <- c", arrowhead="rvee"]
  "c" -- "b" [label="(100) * <- c", arrowhead="rvee"]
  "c" -- "d" [label="(101) * <- c", arrowhead="rvee"]
  "c" -- "e" [label="(102) * <- c", arrowhead="rvee"]
  "c" -- "f" [label="(103) * <- c", arrowhead="rvee"]
  "c" -- "g" [label="(104) * <- c", arrowhead="rvee"]
  "c" -- "h" [label="(105) * <- c", arrowhead="rvee"]
  "c" -- "i" [label="(106) * <- c", arrowhead="rvee"]
  "c" -- "j" [label="(107) * <- c", arrowhead="rvee"]
  "c" -- "k" [label="(108) * <- c", arrowhead="rvee"]
  "b" -- "0" [label="(109) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "a" [label="(110) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "c" [label="(111) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "d" [label="(112) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "e" [label="(113) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "f" [label="(114) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "g" [label="(115) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "h" [label="(116) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "i" [label="(117) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "j" [label="(118) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "b" -- "k" [label="(119) b <-> *", arrowhead="rvee", arrowtail="rvee", dir="both"]
  "d" -- "0" [label="(120) d => *", arrowhead="normal"]
  "d" -- "a" [label="(121) d => *", arrowhead="normal"]
  "d" -- "b" [label="(122) d => *", arrowhead="normal"]
  "d" -- "c" [label="(123) d => *", arrowhead="normal"]
  "d" -- "e" [label="(124) d => *", arrowhead="normal"]
  "d" -- "f" [label="(125) d => *", arrowhead="normal"]
  "d" -- "g" [label="(126) d => *", arrowhead="normal"]
  "d" -- "h" [label="(127) d => *", arrowhead="normal"]
  "d" -- "i" [label="(128) d => *", arrowhead="normal"]
  "d" -- "j" [label="(129) d => *", arrowhead="normal"]
  "d" -- "k" [label="(130) d => *", arrowhead="normal"]
  "e" -- "0" [label="(131) * <= e", arrowhead="normal"]
  "e" -- "a" [label="(132) * <= e", arrowhead="normal"]
  "e" -- "b" [label="(133) * <= e", arrowhead="normal"]
  "e" -- "c" [label="(134) * <= e", arrowhead="normal"]
  "e" -- "d" [label="(135) * <= e", arrowhead="normal"]
  "e" -- "f" [label="(136) * <= e", arrowhead="normal"]
  "e" -- "g" [label="(137) * <= e", arrowhead="normal"]
  "e" -- "h" [label="(138) * <= e", arrowhead="normal"]
  "e" -- "i" [label="(139) * <= e", arrowhead="normal"]
  "e" -- "j" [label="(140) * <= e", arrowhead="normal"]
  "e" -- "k" [label="(141) * <= e", arrowhead="normal"]
  "d" -- "0" [label="(142) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "a" [label="(143) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "b" [label="(144) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "c" [label="(145) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "e" [label="(146) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "f" [label="(147) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "g" [label="(148) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "h" [label="(149) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "i" [label="(150) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "j" [label="(151) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "d" -- "k" [label="(152) d <=> *", arrowhead="normal", arrowtail="normal", dir="both"]
  "f" -- "0" [label="(153) f >> *", style="dashed"]
  "f" -- "a" [label="(154) f >> *", style="dashed"]
  "f" -- "b" [label="(155) f >> *", style="dashed"]
  "f" -- "c" [label="(156) f >> *", style="dashed"]
  "f" -- "d" [label="(157) f >> *", style="dashed"]
  "f" -- "e" [label="(158) f >> *", style="dashed"]
  "f" -- "g" [label="(159) f >> *", style="dashed"]
  "f" -- "h" [label="(160) f >> *", style="dashed"]
  "f" -- "i" [label="(161) f >> *", style="dashed"]
  "f" -- "j" [label="(162) f >> *", style="dashed"]
  "f" -- "k" [label="(163) f >> *", style="dashed"]
  "g" -- "0" [label="(164) * << g", style="dashed"]
  "g" -- "a" [label="(165) * << g", style="dashed"]
  "g" -- "b" [label="(166) * << g", style="dashed"]
  "g" -- "c" [label="(167) * << g", style="dashed"]
  "g" -- "d" [label="(168) * << g", style="dashed"]
  "g" -- "e" [label="(169) * << g", style="dashed"]
  "g" -- "f" [label="(170) * << g", style="dashed"]
  "g" -- "h" [label="(171) * << g", style="dashed"]
  "g" -- "i" [label="(172) * << g", style="dashed"]
  "g" -- "j" [label="(173) * << g", style="dashed"]
  "g" -- "k" [label="(174) * << g", style="dashed"]
  "f" -- "0" [label="(175) f <<>> *", style="dashed", dir="both"]
  "f" -- "a" [label="(176) f <<>> *", style="dashed", dir="both"]
  "f" -- "b" [label="(177) f <<>> *", style="dashed", dir="both"]
  "f" -- "c" [label="(178) f <<>> *", style="dashed", dir="both"]
  "f" -- "d" [label="(179) f <<>> *", style="dashed", dir="both"]
  "f" -- "e" [label="(180) f <<>> *", style="dashed", dir="both"]
  "f" -- "g" [label="(181) f <<>> *", style="dashed", dir="both"]
  "f" -- "h" [label="(182) f <<>> *", style="dashed", dir="both"]
  "f" -- "i" [label="(183) f <<>> *", style="dashed", dir="both"]
  "f" -- "j" [label="(184) f <<>> *", style="dashed", dir="both"]
  "f" -- "k" [label="(185) f <<>> *", style="dashed", dir="both"]
  "h" -- "0" [label="(186) h =>> *"]
  "h" -- "a" [label="(187) h =>> *"]
  "h" -- "b" [label="(188) h =>> *"]
  "h" -- "c" [label="(189) h =>> *"]
  "h" -- "d" [label="(190) h =>> *"]
  "h" -- "e" [label="(191) h =>> *"]
  "h" -- "f" [label="(192) h =>> *"]
  "h" -- "g" [label="(193) h =>> *"]
  "h" -- "i" [label="(194) h =>> *"]
  "h" -- "j" [label="(195) h =>> *"]
  "h" -- "k" [label="(196) h =>> *"]
  "i" -- "0" [label="(197) * <<= i"]
  "i" -- "a" [label="(198) * <<= i"]
  "i" -- "b" [label="(199) * <<= i"]
  "i" -- "c" [label="(200) * <<= i"]
  "i" -- "d" [label="(201) * <<= i"]
  "i" -- "e" [label="(202) * <<= i"]
  "i" -- "f" [label="(203) * <<= i"]
  "i" -- "g" [label="(204) * <<= i"]
  "i" -- "h" [label="(205) * <<= i"]
  "i" -- "j" [label="(206) * <<= i"]
  "i" -- "k" [label="(207) * <<= i"]
  "h" -- "0" [label="(208) h <<=>> *", dir="both"]
  "h" -- "a" [label="(209) h <<=>> *", dir="both"]
  "h" -- "b" [label="(210) h <<=>> *", dir="both"]
  "h" -- "c" [label="(211) h <<=>> *", dir="both"]
  "h" -- "d" [label="(212) h <<=>> *", dir="both"]
  "h" -- "e" [label="(213) h <<=>> *", dir="both"]
  "h" -- "f" [label="(214) h <<=>> *", dir="both"]
  "h" -- "g" [label="(215) h <<=>> *", dir="both"]
  "h" -- "i" [label="(216) h <<=>> *", dir="both"]
  "h" -- "j" [label="(217) h <<=>> *", dir="both"]
  "h" -- "k" [label="(218) h <<=>> *", dir="both"]
  "j" -- "0" [label="(219) j -x *", arrowhead="oinvonormal"]
  "j" -- "a" [label="(220) j -x *", arrowhead="oinvonormal"]
  "j" -- "b" [label="(221) j -x *", arrowhead="oinvonormal"]
  "j" -- "c" [label="(222) j -x *", arrowhead="oinvonormal"]
  "j" -- "d" [label="(223) j -x *", arrowhead="oinvonormal"]
  "j" -- "e" [label="(224) j -x *", arrowhead="oinvonormal"]
  "j" -- "f" [label="(225) j -x *", arrowhead="oinvonormal"]
  "j" -- "g" [label="(226) j -x *", arrowhead="oinvonormal"]
  "j" -- "h" [label="(227) j -x *", arrowhead="oinvonormal"]
  "j" -- "i" [label="(228) j -x *", arrowhead="oinvonormal"]
  "j" -- "k" [label="(229) j -x *", arrowhead="oinvonormal"]
  "k" -- "0" [label="(230) * x- k", arrowhead="oinvonormal"]
  "k" -- "a" [label="(231) * x- k", arrowhead="oinvonormal"]
  "k" -- "b" [label="(232) * x- k", arrowhead="oinvonormal"]
  "k" -- "c" [label="(233) * x- k", arrowhead="oinvonormal"]
  "k" -- "d" [label="(234) * x- k", arrowhead="oinvonormal"]
  "k" -- "e" [label="(235) * x- k", arrowhead="oinvonormal"]
  "k" -- "f" [label="(236) * x- k", arrowhead="oinvonormal"]
  "k" -- "g" [label="(237) * x- k", arrowhead="oinvonormal"]
  "k" -- "h" [label="(238) * x- k", arrowhead="oinvonormal"]
  "k" -- "i" [label="(239) * x- k", arrowhead="oinvonormal"]
  "k" -- "j" [label="(240) * x- k", arrowhead="oinvonormal"]
  "a" -- "0" [label="(241) a :> *", style="bold"]
  "a" -- "b" [label="(242) a :> *", style="bold"]
  "a" -- "c" [label="(243) a :> *", style="bold"]
  "a" -- "d" [label="(244) a :> *", style="bold"]
  "a" -- "e" [label="(245) a :> *", style="bold"]
  "a" -- "f" [label="(246) a :> *", style="bold"]
  "a" -- "g" [label="(247) a :> *", style="bold"]
  "a" -- "h" [label="(248) a :> *", style="bold"]
  "a" -- "i" [label="(249) a :> *", style="bold"]
  "a" -- "j" [label="(250) a :> *", style="bold"]
  "a" -- "k" [label="(251) a :> *", style="bold"]
  "c" -- "0" [label="(252) * <: c", style="bold"]
  "c" -- "a" [label="(253) * <: c", style="bold"]
  "c" -- "b" [label="(254) * <: c", style="bold"]
  "c" -- "d" [label="(255) * <: c", style="bold"]
  "c" -- "e" [label="(256) * <: c", style="bold"]
  "c" -- "f" [label="(257) * <: c", style="bold"]
  "c" -- "g" [label="(258) * <: c", style="bold"]
  "c" -- "h" [label="(259) * <: c", style="bold"]
  "c" -- "i" [label="(260) * <: c", style="bold"]
  "c" -- "j" [label="(261) * <: c", style="bold"]
  "c" -- "k" [label="(262) * <: c", style="bold"]
  "b" -- "0" [label="(263) b <:> *", style="bold", dir="both"]
  "b" -- "a" [label="(264) b <:> *", style="bold", dir="both"]
  "b" -- "c" [label="(265) b <:> *", style="bold", dir="both"]
  "b" -- "d" [label="(266) b <:> *", style="bold", dir="both"]
  "b" -- "e" [label="(267) b <:> *", style="bold", dir="both"]
  "b" -- "f" [label="(268) b <:> *", style="bold", dir="both"]
  "b" -- "g" [label="(269) b <:> *", style="bold", dir="both"]
  "b" -- "h" [label="(270) b <:> *", style="bold", dir="both"]
  "b" -- "i" [label="(271) b <:> *", style="bold", dir="both"]
  "b" -- "j" [label="(272) b <:> *", style="bold", dir="both"]
  "b" -- "k" [label="(273) b <:> *", style="bold", dir="both"]
}"
`;

exports[`render/text/ast2dot #renderAST() - mscgen classic compatible - simple syntax trees should, given a simple syntax tree, render a dot script 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]

  "a" [label="a"];
  "b space" [label="b space"];

  "a" -- "b space" [label="(1) a simple script", arrowhead="normal"]
}"
`;

exports[`render/text/ast2dot #renderAST() - mscgen classic compatible - simple syntax trees should, given a syntax tree with boxes, render a dot script 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]

  "a" [label="a"];
  "b" [label="b"];
  "c" [label="c"];
  "d" [label="d"];

  box1 [shape="note"]
  box1 -- {"a","a"} [style="dotted", dir="none"]
  box2 [shape="box"]
  box2 -- {"b","b"} [style="dotted", dir="none"]
  box3 [shape="hexagon"]
  box3 -- {"c","c"} [style="dotted", dir="none"]
  box4 [style="rounded", shape="box"]
  box4 -- {"d","d"} [style="dotted", dir="none"]
}"
`;

exports[`render/text/ast2dot #renderAST() - mscgen classic compatible - simple syntax trees should, given an 'empty' syntax tree, render a dot script 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]


}"
`;

exports[`render/text/ast2dot #renderAST() - xu compatible alt only - render correct script 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]

  "a" [label="a"];
  "b" [label="b"];
  "c" [label="c"];

  "a" -- "b" [label="(1)", arrowhead="normal"]
  
  subgraph cluster_2{
   label="alt: (2)" labeljust="l"
    "b" -- "c" [label="(3)", arrowhead="normal"]
    "c" -- "b" [label="(4)", style="dashed"]
  }
}"
`;

exports[`render/text/ast2dot #renderAST() - xu compatible alt within loop - render correct script 1`] = `
"/* Sequence chart represented as a directed graph
 * in the graphviz dot language (http://graphviz.org/)
 *
 * Generated by mscgen_js (https://sverweij.github.io/mscgen_js)
 */

graph {
  rankdir=LR
  splines=true
  ordering=out
  fontname="Helvetica"
  fontsize="9"
  node [style=filled, fillcolor=white fontname="Helvetica", fontsize="9" ]
  edge [fontname="Helvetica", fontsize="9", arrowhead=vee, arrowtail=vee, dir=forward]

  "a" [label="a"];
  "b" [label="b"];
  "c" [label="c"];

  "a" -- "b" [label="(1)", arrowhead="normal"]
  
  subgraph cluster_2{
   label="loop: (2) label for loop" labeljust="l"
    
    subgraph cluster_3{
     label="alt: (3) label for alt" labeljust="l"
      "b" -- "c" [label="(4) -> within alt", arrowhead="rvee"]
      "c" -- "b" [label="(5) >> within alt", style="dashed"]
    }
    "b" -- "a" [label="(6) >> within loop", style="dashed"]
  }
  "a" -- "a" [label="(7) happy-the-peppy - outside"]
}"
`;
